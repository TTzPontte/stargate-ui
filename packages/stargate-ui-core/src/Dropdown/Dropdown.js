import React from 'react';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';

import Factory from '../Factory';

const styles = () => {
  return {};
};

const Dropdown = (props) => {
  const { children, ...factoryProps } = props;

  const [classes] = useStyles(styles, {});
  const className = clsx(classes);

  return (
    <Factory className={className} {...factoryProps}>
      {children}
    </Factory>
  );
};

export default Dropdown;
