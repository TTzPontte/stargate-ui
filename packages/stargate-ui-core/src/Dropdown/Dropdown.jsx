import React from 'react';
import PropTypes from 'prop-types';
import uuid from 'react-uuid';

import { useStyles } from '@pontte/stargate-ui-styles';
import Factory from '../Factory';

const styles = (theme) => {
  const {
    palette,
    radius,
    spacing,
  } = theme;

  const dropdown = {
    width: '100%',
    minWidth: 'fit-content',
    maxWidth: 'fit-content',
    lineHeight: 1.1,
    color: palette.darkest,
    border: [1, 'solid', palette.default.lighter.color],
    borderRadius: radius(),
    padding: [spacing(1)],
    cursor: 'pointer',
    display: 'grid',
    alignItems: 'center',
    '& select': {
      '-webkit-appearance': 'none',
      '-moz-appearance': 'none',
      backgroundColor: 'transparent',
      border: 'none',
      padding: [0, '1em', 0, 0],
      margin: [0, '8px', 0, 0],
      width: '100%',
      cursor: 'inherit',
      lineHeight: 'inherit',
      outline: 'none',
      '-webkit-appearance': 'none',
      gridArea: 'select',
    },
      '&:after': {
        pointerEvents: 'none',
        content: '""',
        color: 'black',
        width: '0.8em',
        height: '0.8em',
        border: ['solid', palette.darkest],
        borderWidth: [0, '2px', '2px', 0],
        transform: 'rotate(45deg)',
        borderRadius: radius(.5),
        gridArea: 'select',
        justifySelf: 'end',
        marginLeft: 1,
      },
  }

  return {
    dropdown,
  };
};

const Dropdown = (props) => {
  const {
    options,
    placeholder = 'Selecione',
    disabled,
    value: defaultValue = '',
    onChange = () => {},
    className: inheritedClassName,
    ...factoryProps
  } = props;

  const [value, setValue] = React.useState(defaultValue);

  const handleChange = (event) => {
    if (disabled) {
      event.preventDefault();
      return;
    }

    setValue(event.currentTarget.value);
    onChange(event);
  };

  const [{
    dropdown: classDropdown,
    inheritedClassName,
  }] = useStyles(styles);

  return (
    <Factory className={classDropdown}>
      <Factory
        {...factoryProps}
        key={uuid()}
        disabled={disabled}
        element="select"
        type="select"
        onChange={handleChange}
        value={value}
        aria-label="Dropdown"
      >
        <Factory element="option" value="">{placeholder}</Factory>
        {options.map((option) =>
          <Factory element="option" value={option.value} key={uuid()}>
            {option.label}
          </Factory>
        )}
      </Factory>
    </Factory>
  );
};

Dropdown.displayName = 'Dropdown';

Dropdown.propTypes = {
  /**
   * Disables button and add disabled CSS style.
   *
   * **@default** `undefined`
   */
  disabled: PropTypes.bool,
  /**
   * Value of chosen option.
   *
   * **@default** `''`
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool,
  ]),
  /**
   * Options to select with key: label and value.
   *
   * **@default** `undefined`
   */
  options: PropTypes.array.isRequired,
  /**
   * Placeholder to Dropdown.
   *
   * **@default** `'Selecione'`
   */
  placeholder: PropTypes.string,
  /**
   * Trigger when element is changed.
   * @default Function
   */
  onChange: PropTypes.func,
};

export default Dropdown;
