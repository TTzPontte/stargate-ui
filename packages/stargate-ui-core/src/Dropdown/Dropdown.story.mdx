import {
  Meta,
  Canvas,
  Status,
  Story,
  withAction,
  Props,
} from '@pontte/stargate-ui-playground';

import Dropdown from './Dropdown';

<Meta title="Core/Form Elements/Dropdown" component={Dropdown} />

# Dropdown

<Status type="stable" />

## Basic Usage

<Canvas>
  <Story name="Basic Usage" parameters={{
    jest: ['Dropdown.spec.jsx'],
  }}>
    {
      () => {
        const options = [
          { value: 'one', label: 'One' },
          { value: 'two', label: 'Two' },
          { value: 'three', label: 'Three' },
          { value: 'four', label: 'Four' },
        ];
        return (
          <>
            <Dropdown
              name="dropdownTeste"
              options={options}
              placeholder="Selecione um nÃºmero"
            />
            <Dropdown
              name="dropdownTeste2"
              options={options}
            />
          </>
        );
      }
    }
  </Story>
</Canvas>

<Props component={Dropdown} />

