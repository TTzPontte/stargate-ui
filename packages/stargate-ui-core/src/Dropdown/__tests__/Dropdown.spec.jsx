import React from 'react';

import Dropdown from '../Dropdown';

describe('Dropdown', () => {
  it('should render properly', () => {
    const numbers = [1, 2, 3, 4];
    const { container } = render(
      <Dropdown label="Dropdown" options={numbers} />
    );

    expect(container).toMatchSnapshot();
  });
});
