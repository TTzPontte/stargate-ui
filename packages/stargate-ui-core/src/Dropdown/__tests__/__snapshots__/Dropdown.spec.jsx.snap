// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render properly 1`] = `
<div>
  <div
    class="stargate-ui-factory-0-2-2 stargate-ui-factory-d0-0-2-4 stargate-ui-dropdown-0-2-1"
  >
    <select
      aria-label="Dropdown"
      class="stargate-ui-factory-0-2-6 stargate-ui-factory-d0-0-2-8"
      label="Dropdown"
      type="select"
    >
      <option
        class="stargate-ui-factory-0-2-10 stargate-ui-factory-d0-0-2-12"
        value=""
      >
        Selecione
      </option>
      <option
        class="stargate-ui-factory-0-2-14 stargate-ui-factory-d0-0-2-16"
      />
      <option
        class="stargate-ui-factory-0-2-18 stargate-ui-factory-d0-0-2-20"
      />
      <option
        class="stargate-ui-factory-0-2-22 stargate-ui-factory-d0-0-2-24"
      />
      <option
        class="stargate-ui-factory-0-2-26 stargate-ui-factory-d0-0-2-28"
      />
    </select>
  </div>
</div>
`;
